### ISSUE TITLE
Cursor remains as if suspending on the link after clicking it

### ISSUE THREAD

**Issue by OnceThenEver** [gdh1995/vimium-c#325]
![屏幕截图(253)](https://user-images.githubusercontent.com/66708845/114271876-31462180-9a46-11eb-916d-a49e634ad7dc.png)
![屏幕截图(256)](https://user-images.githubusercontent.com/66708845/114272132-699a2f80-9a47-11eb-8baa-4740714519e8.png)
![屏幕截图(255)](https://user-images.githubusercontent.com/66708845/114271886-3905c600-9a46-11eb-9a42-94d0d652bc2c.png)
↑↑↑↑  New page, new tab ↑↑↑↑

![屏幕截图(254)](https://user-images.githubusercontent.com/66708845/114271885-37d49900-9a46-11eb-8320-3b9dc8c27eea.png)
↑↑↑↑  Then turn back, and the Wikipedia Page Preview emerged (my cursor wasn't on it).   ↑↑↑↑

I have to use _LinkHints.activateModeToLeave_ to remove it.
Nasty, though trivial.

**Comment by gdh1995** [comment]
Does <kbd>Escape</kbd> work and make it hide?

**Comment by OnceThenEver** [comment]
Neither

**Comment by gdh1995** [comment]
The card shows because when you switch back, the page finds the link of `goals` "gets focused again". And this behavior also occurs when I hold on Ctrl+Shift, click the link, move mouse to blank area and switch back.

To work around it, use `map f LinkHints.activate autoUnhover`.

**Comment by gdh1995** [comment]
BTW, `autoUnhover` may prevent LinkHints from making some websites show a card and then clicking it. So if you want you may use a trick to make it only for Wikipedia - https://github.com/gdh1995/vimium-c/wiki/Map-a-key-to-different-commands-on-different-websites .

**Comment by OnceThenEver** [comment]
In the tutorial you provided, `map <c-j:i>` confused me;

> To work around it, use `map f LinkHints.activateMode autoUnhover`.

there are plenty of parameters and commands, like "<c-j:i>" and `autoUnhover`, that I didn't see in any place, including Vimium-C setting page, tutorials in GitHub, and the page you provided in _List of all commands_ at wiki: [https://github.com/gdh1995/vimium-c/blob/master/types/messages.d.ts](url) (abandoned).

So I think there must be more commands that worth leanring, that I didn't know and can't find in normal ways.
I hope that the tutorials at [github.com/gdh1995/vimium-c/wiki](url) will get an augmentation.

**Comment by gdh1995** [comment]
Um I haven't written a help page friendly to non programmers. Up to now, there's only raw source code declaring them.

I've updated the wiki page of https://github.com/gdh1995/vimium-c/wiki/List-of-all-commands , and you can take a look at [types/messages.d.ts#CmdOptions](https://github.com/gdh1995/vimium-c/blob/master/typings/messages.d.ts#:~:text=interface%20CmdOptions) and [background/typed_commands.d.ts](https://github.com/gdh1995/vimium-c/blob/master/background/typed_commands.d.ts).

**Comment by gdh1995** [comment]
Meaning of `<c-j:i>` is in https://github.com/gdh1995/vimium-c/wiki/Use-in-another-keyboard-layout .

**Comment by OnceThenEver** [comment]
`map ' runKey expect={"input":"blank"} keys="<esc>"`
`env input element="input"`

My lines.  ↑
When not focus on the input box I can't use `'` to cancel `enterInsertMode`

`map ' runKey expect={"input":"blank"} keys="<f1>"`
`mapKey <f1> <esc>`

Neither did these.  ↑
But `f1` works in Insert Mode.

**Comment by gdh1995** [comment]
`<esc>` seems not a valid key sequence in your mappings. Here the "keys" field means a valid "key sequence" which has been used in `map xxxKeys xxxCommand`.

The value of `expect` field can be a `Map` of `env-name` and `key-sequence`, so it should not be `blank`.

That is to say, corrected mappings may be:

```
env input element="input"
map ' runKey expect={"input":"<v-blank>"} keys="gF"
map <v-blank> blank
# note: the below "gF" has been mapped to "mainFrame" by default
map gF mainFrame
```

As for `mapKey`, it does not affect `runKey` - it's consumed when recognizing keyboard events - before triggering `map` rules.

Currently there's no command which works exactly as the key `<esc>` does. So why do you want it?

**Comment by OnceThenEver** [comment]
I'd like to make `'` act as `Escape` to exit Insert Mode for that `Escape` is a tiny and marginal key on keyboard, and actually type in character `'` when in input box.

In practice, `mapKey ' <esc>` will always behave like `Escape` whether or not it's in input box.

**Comment by gdh1995** [comment]
Um try `mapKey <':i> <esc>`.

**Comment by gdh1995** [comment]
Sorry, the last commit is wrong. Please try:
```
mapKey ' <esc>
mapKey <':i> '
```

**Comment by gdh1995** [comment]
As for this specific requirement, I have a different idea: I use AutoHotKey to map keys in system level:

``` ahk
!CapsLock:: CapsLock
CapsLock:: Escape
```

The code means `CapsLock` is treated as `Escape`, and `Alt+CapsLock` switches the "Caps" mode.

**Comment by OnceThenEver** [comment]
> ```
> mapKey ' <esc>
> mapKey <':i> '
> ```

These didn't work neither  ↑

> ```
> !CapsLock:: CapsLock
> CapsLock:: Escape
> ```

These were really confusing to me and I pasted them into Shortcut Customization only to get an error message:
"Error: Unknown mapping command: "CapsLock::" in CapsLock:: Escape."

**Comment by gdh1995** [comment]
The second is code for `AutoHotKey`: https://www.autohotkey.com/

**Comment by gdh1995** [comment]
> I'd like to make `'` act as `Escape` to exit Insert Mode for that `Escape` is a tiny and marginal key on keyboard, and actually type in character `'` when in input box.

This sentence itself is inconsistent. When an input box is focused, Vimium C auto enters an Insert Mode.

**Comment by OnceThenEver** [comment]
The Insert Mode I meant is `enterInsertMode`

**Comment by gdh1995** [comment]
`map xxx enterInsertMode key="'"`.

**Comment by OnceThenEver** [comment]
> `map xxx enterInsertMode key="'"`.

Though I still can't figure out what does `key="'"` mean and this line doesn't work,

> Or you may use https://github.com/gdh1995/vimium-c/wiki/Enable-or-Disable-all-frames-by-a-shortcut .

is exactly what I need.

Thank you for your help!

**Comment by gdh1995** [comment]
What I expect is, `key` is an option of the command `enterInsertMode`, and if given, `enterInsertMode` will wait for such a key (in this case, it's `'`) before exiting. But such a `key` option won't take `document.activeElement` into consideration, which means `'` will still exit the global InsertMode even when `<input>` is focused.

As a result, I once added a limit to this option: `its length must > 1` - only those like `<longName>` or `<a-/>` are accepted. This is why `key="'"` doesn't work yet today...

Sorry for my mistake.

Do you think it's a common requirement to allow such printable characters as `enterInsertMode`'s "exiting key"?

**Comment by OnceThenEver** [comment]
Every user of this extension must be one who cannot tolerate the hand movement between mouse and keyboard, though it just take about 0.1s, because even 0.1s is forever long.

I can even describe this fetish as sensitive and idle, but I was not be negative about it, because technology's existence, to some extent, attributes to human's idleness to saving labor. 

So users would like doing all kinds of action with as less hand movements as possible.

I think it's uncommon, really, but when geeks' right forefinger is on `J`, isn't `'` just the nearest key from little finger, if not `;` (like the majority, my handedness is the right)?

**Comment by gdh1995** [comment]
well, but English users might need to type ' many times. This is why I refuse "printable characters" in `key`.




according to what you said, I think it's better to only exclude 0-9, A-Z and a-z. Will it be enough?



---Original---
From: ***@***.***&gt;
Date: Sun, Apr 11, 2021 19:29 PM
To: ***@***.***&gt;;
Cc: "Dahan ***@***.******@***.***&gt;;
Subject: Re: [gdh1995/vimium-c] Cursor remains as if suspending on the link after clicking it (#325)




 
Every user of this extension must be one who cannot tolerate the hand movement between mouse and keyboard, though it just take about 0.1s, because even 0.1s is forever long.
 
I can even describe this fetish as sensitive and idle, but I was not be negative about it, because technology's existence, to some extent, attributes to human's idleness to saving labor.
 
So users would like doing all kinds of action with as less hand movements as possible.
 
I think it's uncommon, really, but when geeks' right forefinger is on J, isn't ' just the nearest key from little finger, if not ; (like the majority, my handedness is the right)?

**Comment by OnceThenEver** [comment]
Add `,` & `.`, they two are also common keys.

**Comment by OnceThenEver** [comment]
In 1.89.1, 
`
map <v-f1> LinkHints.activateMode autoUnhover
map f runKey expect={"youtube":"<v-f2>"} keys="<v-f1>"
`
`f` works at YouTube, but `autoUnhover` in <v-f1> remains as if doesn't exist.

**Comment by gdh1995** [comment]
According to this mapping list, on youtube, there's no autoUnhover - is this just what you want?
On other sites autoUnhover should exist, so where you think it doesn't work? What's the buggy behavior, no hover effects, or Escape not cancelling such effects?



---Original---
From: ***@***.***&gt;
Date: Mon, May 3, 2021 11:54 AM
To: ***@***.***&gt;;
Cc: "Dahan ***@***.******@***.***&gt;;
Subject: Re: [gdh1995/vimium-c] Cursor remains as if suspending on the link after clicking it (#325)




 
In 1.89.1,
 map&nbsp;<v-f1&gt;&nbsp;LinkHints.activateMode&nbsp;autoUnhover map&nbsp;f&nbsp;runKey&nbsp;expect={"youtube":"<v-f2&gt;"}&nbsp;keys="<v-f1&gt;"
 f works at YouTube, but autoUnhover in  remains as if doesn't exist.

**Comment by OnceThenEver** [comment]
The original intention of this issue is to make `LinkHints.activate` not to hover over a link after clicking it, and then you suggested `autoUnhover`.

Now, `autoUnhover` has lost its utility in `<v-f1>`.

`map <v-f2> LinkHints.activate clickable="#thumbnail"`

Why did `<v-f2>` of YouTube work while `<v-f1>` didn't?

**Comment by gdh1995** [comment]
um, what's your full key mapping, including "env"?



---Original---
From: ***@***.***&gt;
Date: Mon, May 3, 2021 12:22 PM
To: ***@***.***&gt;;
Cc: "Dahan ***@***.******@***.***&gt;;
Subject: Re: [gdh1995/vimium-c] Cursor remains as if suspending on the link after clicking it (#325)




 
The original intention of this issue is to make LinkHints.activate not to hover over a link after clicking it, and then you suggested autoUnhover.
 
Now, autoUnhover has lost its utility in <v-f1&gt;.
 
map&nbsp;<v-f2&gt;&nbsp;LinkHints.activate&nbsp;clickable="#thumbnail"
 
Why did <v-f2&gt; of YouTube work while <v-f1&gt; didn't?

**Comment by OnceThenEver** [comment]
```
map <v-f1> LinkHints.activateMode autoUnhover
map <v-f2> LinkHints.activate clickable="#thumbnail"

map f runKey expect={"youtube":"<v-f2>"} keys="<v-f1>"

env youtube host="https://www.youtube.com/"
```

**Comment by OnceThenEver** [comment]
My codes are the same as when in 1.89.0

**Comment by gdh1995** [comment]
Hello, where does your v1.89.1 of Vimium C come from? It should not be released yet, and it seems that I added some mistaken logic into `runKey`.

**Comment by gdh1995** [comment]
Confirmed that `runKey` will directly select the first item in its `expect` option, whatever your `"env"` is.

This bug should only exist in the developing code, but not released versions.

And this can be worked around by adding a new option of `$if={}` to `map ... runKey ....`.

**Comment by OnceThenEver** [comment]
1.89.1 is from #328.

> This bug should only exist in the developing code, but not released versions.

For your words, I'll wait until the next release.
Thank you!

**Comment by OnceThenEver** [comment]
Reopen this issue for one dilemma:
Due to #439, I have to give up using `autoUnhover` to select between choices of the list, but if so, the consequence will be as the above. Even though I may resort to your trick, configuring so many sites is cumbersome and tiring. Being able to cancel the focus by `Esc` after LinkHinting is properly logical, isn't it?

**Comment by OnceThenEver** [comment]
@gdh1995

**Comment by gdh1995** [comment]
Got. I'd like to update `autoUnhover` to skip `<select>`, and make it a boolean or a string of CSS-selector, to allow you to customize its logic.

**Comment by gdh1995** [comment]
I'll update the action fpr long-pressed `<esc>` in normal mode, from `target.blur()` to `unhover(target)` and `target.blur()`. Then you won't need any special parameters - just a repeated `Escape` is enough. If you doesn't like it you may use `autoUnhover="<esc>"` to use a simple `Escape` key.

**Comment by OnceThenEver** [comment]
Solved to the 2nd term of release note 1.95.0
